File: PTB2App/app/src/main/java/com/example/ptb2app/MainActivity.java
// File: MainActivity.java
package com.example.ptb2app;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import java.text.DecimalFormat;

public class MainActivity extends AppCompatActivity {

    EditText edita, editb, editc;
    TextView txtkq;
    Button btnTiepTuc, btnGiai, btnThoat;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        edita = findViewById(R.id.edita);
        editb = findViewById(R.id.editb);
        editc = findViewById(R.id.editc);
        txtkq = findViewById(R.id.txtkq);
        btnTiepTuc = findViewById(R.id.btnTiepTuc);
        btnGiai = findViewById(R.id.btnGiai);
        btnThoat = findViewById(R.id.btnThoat);

        btnGiai.setOnClickListener(v -> {
            String sa = edita.getText().toString();
            String sb = editb.getText().toString();
            String sc = editc.getText().toString();

            if (sa.isEmpty() || sb.isEmpty() || sc.isEmpty()) {
                txtkq.setText(R.string.input_warning);
                return;
            }

            int a = Integer.parseInt(sa);
            int b = Integer.parseInt(sb);
            int c = Integer.parseInt(sc);

            String kq;
            DecimalFormat dcf = new DecimalFormat("0.00");

            if (a == 0) {
                kq = (b == 0) ? (c == 0 ? getString(R.string.infinite_solutions) : getString(R.string.no_solution))
                        : getString(R.string.one_solution, dcf.format(-c / (double) b));
            } else {
                double delta = b * b - 4 * a * c;
                if (delta < 0) {
                    kq = getString(R.string.no_solution);
                } else if (delta == 0) {
                    kq = getString(R.string.double_solution, dcf.format(-b / (2.0 * a)));
                } else {
                    double x1 = (-b + Math.sqrt(delta)) / (2 * a);
                    double x2 = (-b - Math.sqrt(delta)) / (2 * a);
                    kq = getString(R.string.two_solutions, dcf.format(x1), dcf.format(x2));
                }
            }

            txtkq.setText(kq);
        });

        btnTiepTuc.setOnClickListener(v -> {
            edita.setText("");
            editb.setText("");
            editc.setText("");
            txtkq.setText("");
            edita.requestFocus();
        });

        btnThoat.setOnClickListener(v -> finish());
    }
}

File: SumHistoryKotlin_B16/app/src/main/java/com/example/sumhistorykotlin/MainActivity.kt
package com.example.sumhistorykotlin

import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var edtA: EditText
    private lateinit var edtB: EditText
    private lateinit var txtKetQua: TextView
    private lateinit var txtLichSu: TextView
    private lateinit var btnTong: Button
    private lateinit var btnClear: Button

    private val PREF_NAME = "mysave"
    private val KEY_HISTORY = "ls"
    private var lichSu: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)   // dùng layout trực tiếp

        // Ánh xạ view
        edtA = findViewById(R.id.edtA)
        edtB = findViewById(R.id.edtB)
        txtKetQua = findViewById(R.id.txtKetQua)
        txtLichSu = findViewById(R.id.txtLichSu)
        btnTong = findViewById(R.id.btnTong)
        btnClear = findViewById(R.id.btnClear)

        // Load lịch sử đã lưu
        val sp = getSharedPreferences(PREF_NAME, MODE_PRIVATE)
        lichSu = sp.getString(KEY_HISTORY, "") ?: ""
        txtLichSu.text = lichSu

        // Nút TỔNG
        btnTong.setOnClickListener {
            val a = edtA.text.toString().ifBlank { "0" }.toInt()
            val b = edtB.text.toString().ifBlank { "0" }.toInt()
            val kq = a + b
            txtKetQua.text = kq.toString()
            lichSu += "$a + $b = $kq\n"
            txtLichSu.text = lichSu
        }

        // Nút CLEAR
        btnClear.setOnClickListener {
            lichSu = ""
            txtLichSu.text = lichSu
        }
    }

    override fun onPause() {
        super.onPause()
        // Lưu lịch sử
        getSharedPreferences(PREF_NAME, MODE_PRIVATE)
            .edit().putString(KEY_HISTORY, lichSu).apply()
    }
}
